File: Kotlin/test/appendix/ThreadPoolsTests.kt
package appendix

import appendix.ThreadPools.demo1
import appendix.ThreadPools.demo2
import appendix.ThreadPools.demo3
import mocks.out
import java.util.concurrent.CountDownLatch
import java.util.concurrent.Executors
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class ThreadPoolsTests {
  @Test
  fun testThreadPool() {
    out.clear()
    val exec = Executors.newSingleThreadExecutor { Thread(it, "Joe") }
    val end = CountDownLatch(1)
    demo1(exec, end)
    end.await()
    assertEquals("Joe\n", out.toString())
    exec.shutdown()
  }

  @Test
  fun testContext() {
    out.clear()
    val exec = Executors.newSingleThreadExecutor { Thread(it, "Joe") }
    val end = CountDownLatch(1)
    demo2(exec, end)
    end.await()
    assertTrue(out.toString().startsWith("Joe"))
    exec.shutdown()
  }

  @Test
  fun testDispatcher() {
    out.clear()
    val end = CountDownLatch(1)
    demo3(end)
    end.await()
    assertTrue(out.toString().startsWith("DefaultDispatcher"))
  }
}

File: Scala/test/appendix/ThreadPoolsTests.scala
package appendix

import mocks.Appendix.threads
import org.scalatest.funsuite.AnyFunSuite
import tinyscalautils.threads.Executors.global

import java.util.concurrent.CountDownLatch

class ThreadPoolsTests extends AnyFunSuite:
   import ThreadPools.*
   import ThreadPoolsJ.*

   private def runTest(code: CountDownLatch => Unit): Thread =
      val latch = CountDownLatch(1)
      threads.clear()
      code(latch)
      latch.await()
      assert(threads.size == 1)
      threads.remove()

   test("execute (Scala)") {
      val thread = runTest(demo1(global, _))
      assert(thread.getName.startsWith("pool-"))
   }

   test("execute (Java)") {
      val thread = runTest(demo(global, _))
      assert(thread.getName.startsWith("pool-"))
   }

   test("future") {
      val thread = runTest(demo2(global, _))
      assert(thread.getName.startsWith("pool-"))
   }

   test("context") {
      val thread = runTest(demo3)
      assert(thread.getName.contains("-global-"))
   }
